File: pipeline/R/02_features_para_clusters.R
#!/usr/bin/env Rscript

suppressPackageStartupMessages({
  library(tidyverse)
  library(yaml)
  library(arrow)
})

cfg <- yaml::read_yaml("pipeline/config/params.yml")
dir.create(cfg$paths$marts_dir, recursive = TRUE, showWarnings = FALSE)

estab <- arrow::read_parquet(file.path(cfg$paths$curated_dir, "estab_curado.parquet"))
gender <- arrow::read_parquet(file.path(cfg$paths$curated_dir, "gender_curado.parquet"))

nvl <- function(x, default = 0) {
  if (is.null(x)) return(default)
  x
}

build_features <- function(df, level_geo = c("provincia", "departamento"), level_activity = "clae2") {
  level_geo <- match.arg(level_geo)

  geo_cols <- if (level_geo == "provincia") c("provincia_id", "provincia") else c("provincia_id", "provincia", "in_departamentos", "departamento")

  agg <- df |>
    group_by(across(all_of(c("anio", geo_cols, level_activity)))) |>
    summarise(
      empleo = sum(empleo, na.rm = TRUE),
      establecimientos = sum(establecimientos, na.rm = TRUE),
      exportadoras = sum(empresas_exportadoras, na.rm = TRUE),
      .groups = "drop"
    )

  tot <- agg |>
    group_by(across(all_of(c("anio", geo_cols)))) |>
    summarise(
      empleo_total = sum(empleo, na.rm = TRUE),
      estab_total = sum(establecimientos, na.rm = TRUE),
      exp_total = sum(exportadoras, na.rm = TRUE),
      .groups = "drop"
    )

  agg |>
    left_join(tot, by = c("anio", geo_cols)) |>
    mutate(
      share_empleo = if_else(empleo_total > 0, empleo / empleo_total, 0),
      densidad_estab = if_else(empleo > 0, establecimientos / empleo, 0),
      intensidad_export = if_else(establecimientos > 0, exportadoras / establecimientos, 0)
    )
}

for (lg in cfg$run$level_geo) {
  feats <- build_features(estab, level_geo = lg, level_activity = cfg$run$level_activity)
  arrow::write_parquet(feats, file.path(cfg$paths$marts_dir, paste0("features_", lg, ".parquet")))

  if (isTRUE(cfg$run$include_gender_module)) {
    gmodule <- gender |>
      group_by(anio, across(any_of(c("provincia_id", "provincia", "in_departamentos", "departamento", cfg$run$level_activity))), genero) |>
      summarise(empleo = sum(empleo, na.rm = TRUE), .groups = "drop") |>
      tidyr::pivot_wider(names_from = genero, values_from = empleo, values_fill = 0)

    mujeres <- if ("Mujeres" %in% names(gmodule)) gmodule$Mujeres else 0
    varones <- if ("Varones" %in% names(gmodule)) gmodule$Varones else 0

    gmodule <- gmodule |>
      mutate(
        brecha_genero_abs = abs(nvl(mujeres) - nvl(varones)),
        ratio_fem = nvl(mujeres) / pmax(nvl(mujeres) + nvl(varones), 1)
      )

    arrow::write_parquet(gmodule, file.path(cfg$paths$marts_dir, paste0("features_genero_", lg, ".parquet")))
  }
}

message("Features listos en: ", cfg$paths$marts_dir)

File: pipeline/python/02_features_para_clusters.py
#!/usr/bin/env python3
import csv
import json
import os
from collections import defaultdict

CONFIG_PATH = "pipeline/config/params.json"


def load_config(path):
    with open(path, "r", encoding="utf-8") as f:
        return json.load(f)


def read_csv(path):
    with open(path, "r", encoding="utf-8", newline="") as f:
        return list(csv.DictReader(f))


def write_csv(path, rows):
    if not rows:
        raise RuntimeError(f"Sin filas para escribir en {path}")
    with open(path, "w", encoding="utf-8", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=list(rows[0].keys()))
        writer.writeheader()
        writer.writerows(rows)


def to_int(v):
    try:
        return int(float(v))
    except (TypeError, ValueError):
        return None


def to_float(v):
    try:
        return float(v)
    except (TypeError, ValueError):
        return 0.0


def geo_key(row, level_geo):
    if level_geo == "provincia":
        return (to_int(row["provincia_id"]), row["provincia"])
    return (
        to_int(row["provincia_id"]),
        row["provincia"],
        to_int(row["in_departamentos"]),
        row["departamento"],
    )


def build_features(estab_rows, level_geo, level_activity):
    by_cell = defaultdict(lambda: {"empleo": 0.0, "establecimientos": 0.0, "exportadoras": 0.0})
    totals = defaultdict(lambda: {"empleo": 0.0, "establecimientos": 0.0, "exportadoras": 0.0})

    for r in estab_rows:
        g = geo_key(r, level_geo)
        k = (to_int(r["anio"]),) + g + (str(r[level_activity]),)
        by_cell[k]["empleo"] += to_float(r["empleo"])
        by_cell[k]["establecimientos"] += to_float(r["establecimientos"])
        by_cell[k]["exportadoras"] += to_float(r["empresas_exportadoras"])

        tk = (to_int(r["anio"]),) + g
        totals[tk]["empleo"] += to_float(r["empleo"])
        totals[tk]["establecimientos"] += to_float(r["establecimientos"])
        totals[tk]["exportadoras"] += to_float(r["empresas_exportadoras"])

    out = []
    for k, vals in by_cell.items():
        anio = k[0]
        activity = k[-1]
        g = k[1:-1]
        tk = (anio,) + g
        total = totals[tk]

        empleo_total = total["empleo"]
        estab_total = total["establecimientos"]

        row = {
            "anio": anio,
            level_activity: activity,
            "empleo": round(vals["empleo"], 4),
            "establecimientos": round(vals["establecimientos"], 4),
            "exportadoras": round(vals["exportadoras"], 4),
            "empleo_total_geo": round(empleo_total, 4),
            "establecimientos_total_geo": round(estab_total, 4),
            "share_empleo": round(vals["empleo"] / empleo_total, 8) if empleo_total else 0.0,
            "densidad_estab": round(vals["establecimientos"] / vals["empleo"], 8) if vals["empleo"] else 0.0,
            "intensidad_export": round(vals["exportadoras"] / vals["establecimientos"], 8) if vals["establecimientos"] else 0.0
        }

        if level_geo == "provincia":
            row.update({"provincia_id": g[0], "provincia": g[1]})
        else:
            row.update({"provincia_id": g[0], "provincia": g[1], "in_departamentos": g[2], "departamento": g[3]})

        out.append(row)

    return out


def build_gender_features(gender_rows, level_geo, level_activity):
    gsum = defaultdict(lambda: {"mujeres": 0.0, "varones": 0.0, "otro": 0.0})

    for r in gender_rows:
        g = geo_key(r, level_geo)
        k = (to_int(r["anio"]),) + g + (str(r[level_activity]),)
        genero = str(r.get("genero", "")).strip().lower()
        empleo = to_float(r.get("empleo"))

        if genero.startswith("muj"):
            gsum[k]["mujeres"] += empleo
        elif genero.startswith("var"):
            gsum[k]["varones"] += empleo
        else:
            gsum[k]["otro"] += empleo

    out = []
    for k, vals in gsum.items():
        anio = k[0]
        activity = k[-1]
        g = k[1:-1]
        total_bin = vals["mujeres"] + vals["varones"]
        row = {
            "anio": anio,
            level_activity: activity,
            "empleo_mujeres": round(vals["mujeres"], 4),
            "empleo_varones": round(vals["varones"], 4),
            "empleo_otro": round(vals["otro"], 4),
            "ratio_fem": round(vals["mujeres"] / total_bin, 8) if total_bin else 0.0,
            "brecha_genero_abs": round(abs(vals["mujeres"] - vals["varones"]), 4)
        }
        if level_geo == "provincia":
            row.update({"provincia_id": g[0], "provincia": g[1]})
        else:
            row.update({"provincia_id": g[0], "provincia": g[1], "in_departamentos": g[2], "departamento": g[3]})
        out.append(row)
    return out


def export_geo_reference(estab_rows, out_path):
    uniq = {}
    for r in estab_rows:
        key = (r["provincia_id"], r["in_departamentos"])
        uniq[key] = {
            "provincia_id": r["provincia_id"],
            "provincia": r["provincia"],
            "in_departamentos": r["in_departamentos"],
            "departamento": r["departamento"]
        }
    write_csv(out_path, list(uniq.values()))


def main():
    cfg = load_config(CONFIG_PATH)
    os.makedirs(cfg["paths"]["marts_dir"], exist_ok=True)

    estab = read_csv(os.path.join(cfg["paths"]["curated_dir"], "estab_curado.csv"))
    gender = read_csv(os.path.join(cfg["paths"]["curated_dir"], "gender_curado.csv"))

    level_activity = cfg["run"]["level_activity"]

    for level_geo in cfg["run"]["level_geo"]:
        feats = build_features(estab, level_geo, level_activity)
        fpath = os.path.join(cfg["paths"]["marts_dir"], f"features_{level_geo}.csv")
        write_csv(fpath, feats)

        if cfg["run"].get("include_gender_module", False):
            gfeats = build_gender_features(gender, level_geo, level_activity)
            gfpath = os.path.join(cfg["paths"]["marts_dir"], f"features_genero_{level_geo}.csv")
            write_csv(gfpath, gfeats)

    geo_path = os.path.join(cfg["paths"]["marts_dir"], "geo_referencia_para_mapas.csv")
    export_geo_reference(estab, geo_path)
    print("Features OK")


if __name__ == "__main__":
    main()
